lib/plugins/mk-vim-plugin.nix

```
@@ -59,26 +59,74 @@ let
       options = lib.setAttrByPath loc (
         {
           enable = lib.mkEnableOption name;
+          lazyLoad = lib.nixvim.mkLazyLoadOption name;
           autoLoad = lib.nixvim.mkAutoLoadOption cfg name;
         }
         // settingsOption
         // extraOptions
       );
 
       config = lib.mkIf cfg.enable (
+        let
+          globalsConfig = lib.nixvim.applyPrefixToAttrs globalPrefix (cfg.settings or { });
+        in
         lib.mkMerge [
-          {
-            inherit extraPackages extraPlugins;
-            globals = lib.nixvim.applyPrefixToAttrs globalPrefix (cfg.settings or { });
-          }
+          { inherit extraPackages extraPlugins; }
+
+          (lib.mkIf cfg.lazyLoad.enable {
+            assertions = [
+              {
+                assertion = (isColorscheme && colorscheme != null) || cfg.lazyLoad.settings != { };
+                message = "You have enabled lazy loading for ${name} but have not provided any configuration.";
+              }
+            ];
+
+            plugins.lz-n = {
+              plugins = [
+                (
+                  {
+                    # The packpath name is always the derivation name
+                    __unkeyed-1 = lib.getName cfg.package;
+                    # Use provided before, otherwise fallback to normal function wrapped globals config
+                    before =
+                      let
+                        before = cfg.lazyLoad.settings.before or null;
+                        default = lib.mkIf (lib.nixvim.toLuaObject globalsConfig != "{ }") ''
+                          function()
+                            local globals = ${lib.nixvim.toLuaObject globalsConfig}
+
+                            for k,v in pairs(globals) do
+                              vim.g[k] = v
+                            end
+                          end
+                        '';
+                      in
+                      if (lib.isString before || lib.types.rawLua.check before) then before else default;
```

In nix, we don't need parentheses for `if` conditions

```suggestion
                      if lib.isString before || lib.types.rawLua.check before then before else default;
```

---

Thanks for reminding me of the history.

There's no "one right approach" here:
- Based on a freeform perspective, the attr can be anything so a stricter type check makes sense.
- If there's an actual option declared, though, it probably makes sense to make its effective default its actual default.

I agree, however we handle it, it should be consistent.

I'm fine for this to be merged as is, so we can consider refractors to both implementations in a dedicated PR.